api/client/client_test.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+)
+
+// mockServer mocks an Auth Server.
+type mockServer struct {
+	grpc *grpc.Server
+	*proto.UnimplementedAuthServiceServer
+}
+
+func newMockServer() *mockServer {
+	m := &mockServer{
+		grpc.NewServer(),
+		&proto.UnimplementedAuthServiceServer{},
+	}
+	proto.RegisterAuthServiceServer(m.grpc, m)
+	return m
+}
+
+func (m *mockServer) Ping(ctx context.Context, req *proto.PingRequest) (*proto.PingResponse, error) {
+	return &proto.PingResponse{}, nil
+}
+
+// mockInsecureCredentials mocks insecure Client credentials.
+// it returns a nil tlsConfig which allows the client to run in insecure mode.
+type mockInsecureTLSCredentials struct{}
+
+func (mc *mockInsecureTLSCredentials) Dialer(cfg Config) (ContextDialer, error) {
+	return nil, trace.NotImplemented("no dialer")
+}
+
+func (mc *mockInsecureTLSCredentials) TLSConfig() (*tls.Config, error) {
+	return nil, nil
+}
+
+func (mc *mockInsecureTLSCredentials) SSHClientConfig() (*ssh.ClientConfig, error) {
+	return nil, trace.NotImplemented("no ssh config")
+}
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	addr := "localhost:3025"
+	startMockServer(t, addr)
+
+	tests := []struct {
+		desc      string
+		config    Config
+		assertErr require.ErrorAssertionFunc
+	}{{
+		desc: "successfully dial tcp address.",
+		config: Config{
+			Addrs: []string{addr},
+			Credentials: []Credentials{
+				&mockInsecureTLSCredentials{},
+			},
+			DialOpts: []grpc.DialOption{
+				grpc.WithInsecure(),
```

Comment:
Leave a TODO notice to get rid of Insecure in these tests here and everywhere below where it's used.